src/clj_foundation/grep.clj

```diff
@@ -0,0 +1,102 @@
+(ns clj-foundation.grep
+  "Unix-style grep for arbitrary Clojure nested collections."
+  (:require [clojure.zip :as zip]
+            [clojure.set :as set]
+            [clj-foundation.tree-visit :as v]
+            [clj-foundation.patterns :refer [nothing]]))
+
+
+(defn- update-breadcrumb [old-path loc]
+  (let [parent (v/parent-container loc)
+        new-depth (v/depth loc)
+        new-index (v/index loc)
+        new-node (zip/node loc)]
+    (if (or (map? new-node) (v/root-node? (zip/up loc)))
+      old-path
+      (if (map? parent)
+        (if (not (v/traversable-coll? new-node))
+          (conj (vec (take (- new-depth 1) old-path)) new-node)
+          old-path)
+        (conj (vec (take (- new-depth 1) old-path)) new-index)))))
+
+
+(defn- found-match-container
+  "Add a result to state containing the breadcrumb and the parent container of the matched node."
+  [state breadcrumb loc]
+  {:state (conj state [breadcrumb (v/parent-container loc)])})
+
+
+(defn- found-match-node
+  "Add a result to state containing the breadcrumb and the clojure.zip visitor node referencing
+  the match."
+  [state breadcrumb loc]
+  {:state (conj state [breadcrumb loc])})
+
+
+(defn- grep-match?
+  "Is the current node a match?  If so, capture it in state using found-match-fn."
+  [pattern node state loc breadcrumb found-match-fn]
+  (if-not (or (v/traversable-coll? node) (nil? node))
+    (cond
+      (and (instance? java.util.regex.Pattern pattern) (re-matches pattern (.toString node))) (found-match-fn state breadcrumb loc)
```

Comment:
I might default to "re-find" or add an option to use re-find, as it is more "greppy"